Rewrite the introduction (#510)
JuliaDiff/ChainRulesCore.jl: docs/src/index.md

# ChainRules

[Automatic differentiation (AD)](https://en.wikipedia.org/wiki/Automatic_differentiation) is a set of techniques for obtaining derivatives of arbitrary functions.

> There is a surprisingly large number of packages for doing AD in Julia.

Reviewer: ```suggestion
There are a surprisingly large number of packages for doing AD in Julia.
```

> ChainRules isn't one of these packageis.

Author: ```suggestion
ChainRules isn't one of these packages.
```

Reviewer: I love this pair of sentences

The AD packages essentially combine derivatives of simple functions into derivatives of more complicated functions.

> They differ in the way they combine the derivatives, but they all require a common set of derivatives of simple functions (primitives/rules).

Reviewer: the difference is less  in how they combine derivatives and more in how they break down to the simple functions, IMO.
But that might be too much detail.

They should also have a test-only dependency on [ChainRulesTestUtils.jl](https://github.com/JuliaDiff/ChainRulesTestUtils.jl) to test the rules using finite differences.

Note that the packages with rules do not have to depend on AD systems, and neither do the AD systems have to depend on individual packages.

## Key functionality

Consider a relationship $y = f(x)$, where $f$ is some function.
Computing $y$ from $x$ is the original problem, called the _primal_ computation, in contrast to the problem of computing derivatives.
We say that the _primal function_ $f$ takes a _primal input_ $x$ and returns the _primal output_ $y$.

ChainRules rules are concerned with propagating _tangents_ of primal inputs to _tangents_ of primal outputs (`frule`, from forwards mode AD), and propagating _cotangents_ of primal outputs to _cotangents_ of primal inputs (`rrule`, from reverse mode AD).
To be able to do that, ChainRules also defines a small number of tangent types to represent tangents and cotangents.

!!! note "Tangents and cotangents"
    Strictly speaking tangents, $ẋ = \frac{dx}{da}$, are propagated in `frule`s, and cotangents, $x̄ = \frac{da}{dx}$, are propagated in `rrule`s.

>     However, we do not distinguish between the two in practice: both are represented by the same tangent types, and we sometimes (often?) refer to cotangents as tangents.

Reviewer: ```suggestion
    However, in practice there is rarely a need to distinguish the two: both are represented by the same tangent types.
    Thus, except when the detail might clarify, we refer to both as tangents.
```

Reviewer 2: Tangents and cotangents need a proper, accessible introduction imo. Can explain it first as "propagating derivatives" of some "outermost function" `F` wrt the inputs of a given function `g` to the derivatives of this outermost function `F` wrt the outputs of `g` (frule) or the other way around (rrule).

Author: Yes, I think we should have that in the maths section. I will open an issue, it's out of scope of rewriting the intro IMO.

!!! terminology "`frule` and `rrule`"

>     The whole field is a mess for terminology.

Reviewer: This should be the last sentence i think

### Tangent types

The types of (co)-tangents depend on the types of the primals.
Scalar primals are represented by scalar tangents (e.g. `Float64` tangent for a `Float64` primal).
Vector, matrix, and higher rank tensor primals can be represented by vector, matrix and tensor tangents.

ChainRules defines a [`Tangent`](@ref) tangent type to represent tangents of `struct`s, `Tuple`s, `NamedTuple`s, and `Dict`s.

Additionally, for signalling semantics, we distinguish between two tangent types representing a zero tangent.
[`NoTangent`](@ref) type represent situtations in which the tangent space does not exist, e.g. an index into an array can not be perturbed.
[`ZeroTangent`](@ref) is used for cases where the tangent happens to be zero, e.g. because the primal argument is not used in the computation.

We also define [`Thunk`](@ref)s to allow certain optimisation.
`Thunk`s are a wrapper over a computation that can potentially be avoided, depending on the downstream use.

See the section on [tangent types](@ref tangents) for more details.

> ## Videos

Reviewer: Can we push these right down to the bottom, or even back into their own page?
